lib/irb/type_completion/scope.rb

```diff
@@ -0,0 +1,408 @@
+# frozen_string_literal: true
+
+require 'set'
+require_relative 'types'
+module IRB
+  module TypeCompletion
+    class BaseScope
```

Comment:
Why do we need to extract `BaseScope` out of `Scope`?

Reply:
Scope represents mutable virtual scope that has table of what variable will be changed during type analysis.
BaseScope represents immutable existing scope in current binding. It does not have table of variable changes. It just retrieves the actual value in current binding.

`base_scope['lvar']` will execute
```ruby
binding.eval('lvar') if binding.local_variables.include?(:lvar)
```

scope['lvar'] will execute
```ruby
if @table.has? 'lvar'
  # When lvar is changed in the input, (example: `lvar = lvar.foo; lvar.`)
  # Type of lvar.foo is stored in @table['lvar']
  @table['lvar']
else
  parent_scope['lvar'] # parent_scope is base_scope or another scope. root scope is always base_scope
end
```

Reply:
It seems that `Scope` actually doesn't share many ivars with `BaseScope`. It also doesn't utilise `BaseScope`'s implementation much (e.g. calling `super`). Furthermore, a `Scope` always has a parent while `BaseScope` doesn't. The things they mostly share seem to be constants like `BREAK_RESULT`?

These make me wonder if using inheritance here is the best option. They seem to be very different structure that just respond to the same interfaces. If that's the case, I think `Scope` shouldn't inherit `BaseScope` as it'd cause a wrong impression to readers.

Perhaps renaming them to `RootScope` and `LeafScope` will also help clarify their relationships.

Reply:
I see. Yes, it's not a inheritance. RootScope name looks good.
I think LeafScope is not accurate because it can be a parent of another LeafScope, so I'll keep the name Scope.